### DISCUSSION TITLE
Handling missing packages in specific functionalities

### DISCUSSION THREAD

**Discussion by leeuwis** [instantcommerce/next-api-decorators#115] [General]
The following functionalities need specific packages to be installed:
- Data Transfer Objects: [validateObject](https://github.com/storyofams/next-api-decorators/blob/e68c91601a02d73f0ed5b633fc26bc109d9c68d8/lib/internals/validateObject.ts#L7) requires `class-validator` and `class-transformer` (if `ValidationPipe` is used)
- Route Matching: [findRoute](https://github.com/storyofams/next-api-decorators/blob/master/lib/internals/findRoute.ts) requires `path-to-regexp` (if non-empty paths are used in the method decorators)

At the moment this is not handled explicitly, resulting in unexpected behaviour (if someone didn't fully read the docs).

Should we:
- Install the packages by default (makes is less "light", but does that matter since it's server side)
- Throw an error (crash) when one of the functions is used without the package(s)
- Display a warning with a link to the corresponding docs when one of the functions is used without the package(s)

Let me know your thoughts 🤙

**Comment by ggurkal** [comment]
I think we should display a warning message. About the other options:

> Install the packages by default

We shouldn't install packages by default because it'll have an impact on the cold start of lambda functions.

> Throw an error (crash) when one of the functions is used without the package(s)

If we crash, the function will go into a loop of restart -> crash -> restart I believe.

**Reply by leeuwis** [reply]
@ggurkal implemented the warnings in [this PR](https://github.com/storyofams/next-api-decorators/pull/116).

Regarding the error option, we could throw an `InternalServerErrorException` in the `validateObject` handler. This wouldn't result in a crash loop, it would return a HTTP exception when a route using the `ValidationPipe` is called. What are your thoughts on this?

**Reply by ggurkal** [reply]
I've done some checks and the following is only valid for `class-validator` (and maybe `class-transformer`):

This functionality will be useful when a dependency is in `devDependencies` rather than `dependencies`, in that case build process will succeed but the production build won't even start.

So, my question is: does it make sense to add some extra functionality for development mode, like check if the package is in `devDependencies` while it should be in `dependencies` and warn the dev as early as possible?

If you want to test, move `class-validator` into `devDependencies` and:
```bash
$ yarn build && rm -rf node_modules && yarn --production && yarn start
```
